SWDEV-546350 - [catch2][dtest] Test cases for hipMemPrefetchAsync_v2 and hipMemAdvise_v2

swargamrambabu opened the pull request:
## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->
SWDEV-546350 : To Test the basic functionality of hipMemPrefetchAsync_v2 and hipMemAdvise_v2 API's

## Technical Details

<!-- Explain the changes along with any relevant GitHub links. -->
hipMemPrefetchAsync_v2 and hipMemAdvise_v2 API's Test cases

## Test Plan

<!-- Explain any relevant testing done to verify this PR. -->
Adding hip-tests with this PR

## Test Result

<!-- Briefly summarize test outcomes. -->
All tests are passed internally

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

swargamrambabu committed:
SWDEV-546350 - [catch2][dtest] Test cases for hipMemPrefetchAsync_v2 and hipMemAdvice_v2

cjatin commented on projects/hip-tests/catch/unit/memory/hipMemAdvice_v2.cc:
```diff
@@ -0,0 +1,222 @@
+  int* memPtr = nullptr;
+
+  hipStream_t stream;
+  HIP_CHECK(hipStreamCreate(&stream));
```
where is this stream used?

swargamrambabu replied:
Removed

cjatin commented on projects/hip-tests/catch/unit/memory/hipMemAdvice_v2.cc:
```diff
@@ -0,0 +1,222 @@
+  const int N = 16;
+  const int Nbytes = N * sizeof(int);
+
+  hipStream_t stream;
```
same question, where is this stream used?

swargamrambabu replied:
Removed

swargamrambabu committed:
SWDEV-546350 - Enable tests for NVIDIA, Guard failing tests

rahulc-gh commented:
/AzurePipelines run rocm-ci-caller

swargamrambabu commented:
!verify

swargamrambabu committed:
SWDEV-546350 - Add checks for Managed Memory Support

swargamrambabu committed:
SWDEV-546350 - Gaurded Numa related things against windows

swargamrambabu commented:
/AzurePipelines run rocm-ci-caller

sourabhuday commented:
Can #if 0/ #endif guards be removed in this PR and added back once those issues are fixed?

swargamrambabu committed:
SWDEV-546350 - Removed some scenarios which needs fixes, gaurded for NVIDIA

swargamrambabu commented:
> Can #if 0/ #endif guards be removed in this PR and added back once those issues are fixed?

Hi @sourabhuday 
Removed some scenarios which needed fix and guarded against NVIDIA directly in Cmake file to avoid compilation issues.
Could you please check once. 
We will add those scenarios back in new PR once the fixes for SWDEV-551244 , SWDEV-551254, SWDEV-551259 are available

cjatin commented on projects/hip-tests/catch/unit/memory/hipMemAdvice_v2.cc:
```diff
@@ -0,0 +1,292 @@
+ * Kernel to fill value for each element in the given array
+ */
+static __global__ void fillDataKernel(int *arr, size_t size, int value) {
+  size_t offset = blockDim.x * blockIdx.x + threadIdx.x;
```
You do not need this sort offset based indexing for a 1D array, can we simplify it.

swargamrambabu replied:
Modified, simply used threadIdx.x 
Also modified kernel to copy data

cjatin commented on projects/hip-tests/catch/unit/memory/hipMemAdvice_v2.cc:
```diff
@@ -0,0 +1,292 @@
+
+  HIP_CHECK(hipSetDevice(supportedDevices[0]));
+
+  if (numa_available() < 0) {
```
can we move this to line 149.
`if (supportedDevices.empty() || numa_available() < 0)`

swargamrambabu replied:
Done

cjatin commented on projects/hip-tests/catch/unit/memory/hipMemAdvice_v2.cc:
```diff
@@ -0,0 +1,292 @@
+      HIP_CHECK(
+          hipMemAdvise_v2(memPtr, Nbytes, hipMemAdviseSetReadMostly, location));
+
+      for (int i = 0; i < N; i++) {
```
we are writing this memory again and again for each node, should not this be written once and then do a MemAdvise call?

swargamrambabu replied:
Modified. Written initially only.
And modified all scenarios only to read after advice, avoided writing data after advice call with hipMemAdviseSetReadMostly

cjatin commented on projects/hip-tests/catch/unit/memory/hipMemPrefetchAsync_v2.cc:
```diff
@@ -0,0 +1,306 @@
+ * Kernel to fill value for each element in the given array
+ */
+static __global__ void fillDataKernel(int *arr, size_t size, int value) {
+  size_t offset = blockDim.x * blockIdx.x + threadIdx.x;
```
For 1D array, can we simplify this indexing.

swargamrambabu replied:
Modified, simply used threadIdx.x 
Also added new kernel to copy data

cjatin commented on projects/hip-tests/catch/unit/memory/hipMemPrefetchAsync_v2.cc:
```diff
@@ -0,0 +1,306 @@
+
+  SECTION("With Device") {
+    int deviceCount = 0;
+    HIP_CHECK(hipGetDeviceCount(&deviceCount));
```
Where is this deviceCount used?

swargamrambabu replied:
removed

cjatin commented on projects/hip-tests/catch/unit/memory/hipMemPrefetchAsync_v2.cc:
```diff
@@ -0,0 +1,306 @@
+    hipMemLocation location;
+    location.type = hipMemLocationTypeHost;
+
+    HIP_CHECK(hipMemPrefetchAsync_v2(memPtr, Nbytes, location, 0, stream));
```
I am confused about this test, we prefetch async to host first then write to it? we never read it.

swargamrambabu replied:
Initially only tried to directly written data to memory after prefetch and  validated.
In the latest code modified as,
1) Create memory
2) Fill the data
3) Prefetch
4) Get and validate it by copying 
5) Fill with new data
6) Validate

cjatin replied:
still a bit confused. We filled the data via host and prefetch it to host?

swargamrambabu replied:
For the device scenarios, at start filled data through host. And for Host, Host Numa, Host Numa current scenarios filled data through device at start

swargamrambabu replied:
For the device scenarios, at start filled data through host. And for Host, Host Numa, Host Numa current scenarios filled data through device at start

swargamrambabu replied:
For the device scenarios, at start filled data through host. And for Host, Host Numa, Host Numa current scenarios filled data through device at start

cjatin commented on projects/hip-tests/catch/unit/memory/hipMemPrefetchAsync_v2.cc:
```diff
@@ -0,0 +1,306 @@
+      location.type = hipMemLocationTypeHostNuma;
+      location.id = node;
+
+      HIP_CHECK(hipMemPrefetchAsync_v2(memPtr, Nbytes, location, 0, stream));
```
Same question, we prefetch it first and then write it?

swargamrambabu replied:
Modified

swargamrambabu committed:
SWDEV-546350 - Incorporated review comments

swargamrambabu commented:
Hi @cjatin 
Addressed the review comments, could you please check once.

swargamrambabu committed:
SWDEV-546350 - Corrected File name

cjatin commented on projects/hip-tests/catch/unit/memory/hipMemAdvise_v2.cc:
```diff
@@ -0,0 +1,278 @@
+
+  SECTION("With Device") {
+    for (int deviceId : supportedDevices) {
+      HIP_CHECK(hipSetDevice(deviceId));
```
After this setdevice, at the end of the test we should reset it to default device. Other tests might expect a particular device to be set.

swargamrambabu replied:
At end of test re-setted to device 0

cjatin commented on projects/hip-tests/catch/unit/memory/hipMemAdvise_v2.cc:
```diff
@@ -0,0 +1,278 @@
+                           "supported devices or No Numa nodes in system");
+  }
+
+  HIP_CHECK(hipSetDevice(supportedDevices[0]));
```
same comment here, after the test is done, reset it to original device (which will be 0 i guess)

swargamrambabu replied:
At end of test re-setted to device 0 for all the tests

swargamrambabu committed:
SWDEV-546350 - Resetted to default device, modifed initial data filling ways

swargamrambabu commented:
Hi @cjatin 
Addressed the review comments, could you please check once.

swargamrambabu committed:
Merge branch 'develop' into amd/dev/rswargam/SWDEV-546350

jainprad committed:
Merge branch 'develop' into amd/dev/rswargam/SWDEV-546350

jainprad committed:
Merge branch 'develop' into amd/dev/rswargam/SWDEV-546350

jainprad committed:
Merge branch 'develop' into amd/dev/rswargam/SWDEV-546350

rakesroy commented:
Stress test passed for PSDB build Linux# 2979, Windows# 875

jainprad committed:
Merge branch 'develop' into amd/dev/rswargam/SWDEV-546350

jainprad commented:
/AzurePipelines run rocm-ci-caller
